interchain-token-service/src/contract/execute.rs

```diff
@@ -0,0 +1,291 @@
+use cosmwasm_std::{DepsMut, HexBinary, Response};
+use error_stack::{report, Result, ResultExt};
+use router_api::{Address, ChainName, CrossChainId};
+
+use crate::contract::Error;
+use crate::events::ItsContractEvent;
+use crate::primitives::ItsHubMessage;
+use crate::state::{self, load_config, load_its_address};
+
+/// Executes an incoming ITS message.
+///
+/// This function handles the execution of ITS (Interchain Token Service) messages received from
+/// its sources. It verifies the source address, decodes the message, applies balance tracking,
+/// and forwards the message to the destination chain.
+pub fn execute_message(
+    deps: DepsMut,
+    cc_id: CrossChainId,
+    source_address: Address,
+    payload: HexBinary,
+) -> Result<Response, Error> {
+    let config = load_config(deps.storage).change_context(Error::InvalidStoreAccess)?;
+
+    let source_chain = ChainName::try_from(cc_id.source_chain.clone().to_string())
+        .change_context(Error::InvalidPayload)?;
+    let its_source_address =
+        load_its_address(deps.storage, &source_chain).change_context(Error::InvalidStoreAccess)?;
+    if source_address != its_source_address {
+        return Err(report!(Error::UnknownItsAddress(source_address)));
+    }
+
+    let its_hub_message =
+        ItsHubMessage::abi_decode(&payload).change_context(Error::InvalidPayload)?;
+
+    match its_hub_message {
+        ItsHubMessage::SendToHub {
+            destination_chain,
+            message: its_message,
+        } => {
+            let destination_payload = ItsHubMessage::ReceiveFromHub {
+                source_chain: source_chain.clone(),
+                message: its_message.clone(),
+            }
+            .abi_encode();
+
+            let destination_address = load_its_address(deps.storage, &destination_chain)
+                .change_context(Error::InvalidStoreAccess)?;
```

Comment:
Is this safe? It seems a little weird that when someone creates the original message, they don't necessarily know the final destination address for the message, but rather it's supplied by the ITS hub. And it looks like this is modifiable? So between when I send the message and when it actually arrives at the ITS hub, the destination ITS contract could change. In what situation would the ITS contract for a particular chain change?

Reply:
ITS hub is a router for ITS calls. Similar to how external gateway contracts don't know about each other's address, since the Amplifier protocol does the translation, there's no need for each ITS edge contract to also be aware. In fact, this is one of the powerful properties of our hub design that we want to extend to ITS as well as we move away from the p2p configuration.

Reply:
Yeah, I'm also considering preventing the overwrite for it and making it require an explicit migration like for gateway address. It's unlikely we redeploy ITS contracts, and would require a migration externally as well

Reply:
Thought about allowing overwrite more. Unlike external gateway addresses that are maintained per voting verifier, ITS hub has to track all ITS addresses. So if an integrator whitelists an ITS address incorrectly, or needs to redeploy (quite probable during testing), it'll be convenient to allow overwriting, so we don't have to migrate ITS hub